Source/Meadow.Foundation.Peripherals/Sensors.Gnss.Bg95M3/Driver/Bg95M3.cs

```diff
@@ -0,0 +1,182 @@
+﻿using System;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Meadow.Foundation.Sensors.Location.Gnss;
+using Meadow.Hardware;
+using Meadow.Peripherals.Sensors.Location.Gnss;
+
+namespace Meadow.Foundation.Sensors.Gnss.Bg95M3
+{
+    /// <summary>
+    /// Represents a BG95-M3 Cellular/GNSS module
+    /// </summary>
+    public class Bg95M3 : IGnssSensor
+    {
+        private bool isGnssUpdateEnabled;
+
+        private TimeSpan _updatePeriod = TimeSpan.FromMinutes(30);
+
+        private ICellNetworkAdapter _cellAdapter;
+
+        private CancellationTokenSource cancellationTokenSource;
+
+        private NmeaSentenceProcessor nmeaProcessor;
+        
+        private IGnssResult[] _supportedResultTypes = new IGnssResult[]
+        {
+            new GnssPositionInfo(),
+            new ActiveSatellites(),
+            new CourseOverGround(),
+            new SatellitesInView(new Satellite[0]),
+        };
+
+        /// <summary>
+        /// Initializes NMEA sentence decoders for GNSS data processing based on supported result types
+        /// </summary>
+        private void InitDecoders()
+        {
+            nmeaProcessor = new NmeaSentenceProcessor();
+
+            foreach (var resultType in SupportedResultTypes)
+            {
+                if (resultType is GnssPositionInfo)
+                {
+                    var ggaDecoder = new GgaDecoder();
+                    nmeaProcessor.RegisterDecoder(ggaDecoder);
+                    ggaDecoder.PositionReceived += (sender, location) =>
+                    {
+                        GnssDataReceived(this, location);
```

Comment:
I tend to use the ?.Invoke pattern just be sure the EventHandler is correctly set